File: frontend/js/caixa.js
function formatarMoedaCaixa(valor) {
  return Number(valor || 0).toLocaleString("pt-BR", { style: "currency", currency: "BRL" });
}

const NOMES_FORMA_PAGAMENTO = {
  dinheiro: "Dinheiro",
  pix: "PIX",
  cartao_debito: "Cartão de débito",
  cartao_credito: "Cartão de crédito",
  fiado: "Fiado",
};

async function carregarCaixaAtual() {
  const sessao = await Api.caixaAtual();

  document.getElementById("bloco-caixa-fechado").classList.toggle("oculto", !!sessao);
  document.getElementById("bloco-caixa-aberto").classList.toggle("oculto", !sessao);

  if (!sessao) return;

  document.getElementById("valor-abertura-atual").textContent = formatarMoedaCaixa(sessao.valor_abertura);
  document.getElementById("valor-total-vendas").textContent = formatarMoedaCaixa(sessao.resumo.total_vendas);
  document.getElementById("valor-suprimentos").textContent = formatarMoedaCaixa(sessao.resumo.suprimentos);
  document.getElementById("valor-sangrias").textContent = formatarMoedaCaixa(sessao.resumo.sangrias);
  document.getElementById("valor-esperado-dinheiro").textContent = formatarMoedaCaixa(sessao.resumo.valor_esperado_dinheiro);

  const porForma = sessao.resumo.vendas_por_forma_pagamento;
  document.getElementById("tabela-vendas-forma").innerHTML =
    Object.entries(porForma)
      .map(([forma, total]) => `<tr><td>${NOMES_FORMA_PAGAMENTO[forma] || forma}</td><td>${formatarMoedaCaixa(total)}</td></tr>`)
      .join("") || "<tr><td colspan='2'>Nenhuma venda nesta sessão ainda.</td></tr>";

  await carregarMovimentacoesCaixa();
}

async function carregarMovimentacoesCaixa() {
  const movimentacoes = await Api.listarMovimentacoesCaixa();
  document.getElementById("tabela-movimentacoes-caixa").innerHTML =
    movimentacoes
      .map(
        (m) => `
          <tr>
            <td>${new Date(m.criado_em).toLocaleString("pt-BR")}</td>
            <td><span class="selo ${m.tipo === "suprimento" ? "selo-ativo" : "selo-inativo"}">${m.tipo}</span></td>
            <td>${formatarMoedaCaixa(m.valor)}</td>
            <td>${m.motivo || "—"}</td>
            <td>${m.usuario_nome || "—"}</td>
          </tr>
        `
      )
      .join("") || "<tr><td colspan='5'>Nenhuma movimentação registrada nesta sessão.</td></tr>";
}

async function carregarSessoesCaixa() {
  const sessoes = await Api.listarSessoesCaixa();
  document.getElementById("tabela-sessoes-caixa").innerHTML =
    sessoes
      .map(
        (s) => `
          <tr>
            <td>${new Date(s.aberto_em).toLocaleString("pt-BR")}</td>
            <td>${s.fechado_em ? new Date(s.fechado_em).toLocaleString("pt-BR") : "—"}</td>
            <td>${formatarMoedaCaixa(s.valor_abertura)}</td>
            <td>${s.valor_fechamento_informado !== null ? formatarMoedaCaixa(s.valor_fechamento_informado) : "—"}</td>
            <td><span class="selo ${s.status === "aberto" ? "selo-ativo" : "selo-info"}">${s.status}</span></td>
          </tr>
        `
      )
      .join("") || "<tr><td colspan='5'>Nenhuma sessão de caixa registrada ainda.</td></tr>";
}

document.getElementById("btn-abrir-caixa").addEventListener("click", async () => {
  await Api.abrirCaixa(document.getElementById("valor-abertura").value || 0);
  await carregarCaixaAtual();
  await carregarSessoesCaixa();
});

document.getElementById("btn-registrar-movimentacao-caixa").addEventListener("click", async () => {
  const valor = document.getElementById("movimentacao-caixa-valor").value;
  if (!valor || Number(valor) <= 0) {
    alert("Informe um valor maior que zero.");
    return;
  }
  try {
    await Api.criarMovimentacaoCaixa({
      tipo: document.getElementById("movimentacao-caixa-tipo").value,
      valor,
      motivo: document.getElementById("movimentacao-caixa-motivo").value,
    });
    document.getElementById("movimentacao-caixa-valor").value = "";
    document.getElementById("movimentacao-caixa-motivo").value = "";
    await carregarCaixaAtual();
  } catch (erro) {
    alert(erro.message);
  }
});

document.getElementById("btn-fechar-caixa").addEventListener("click", async () => {
  if (!confirm("Fechar o caixa? Isso encerra a sessão atual.")) return;
  await Api.fecharCaixa({
    valor_fechamento_informado: document.getElementById("valor-fechamento").value || null,
    observacoes: document.getElementById("fechamento-observacoes").value,
  });
  await carregarCaixaAtual();
  await carregarSessoesCaixa();
});

document.addEventListener("DOMContentLoaded", async () => {
  await carregarCaixaAtual();
  await carregarSessoesCaixa();
});

File: backend/app/routes/caixa.py
from datetime import datetime, timezone

from flask import Blueprint, request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity

from app.extensions import db
from app.models.caixa import CaixaSessao, CaixaMovimentacao, TIPOS_MOVIMENTACAO_CAIXA_VALIDOS
from app.models.venda import Venda
from app.utils.rbac import perfis_permitidos
from app.utils.tenant import query_tenant, loja_atual_id

caixa_bp = Blueprint("caixa", __name__)
_PDV = ("admin", "gerente", "vendedor")


def _sessao_aberta(loja_id):
    return CaixaSessao.query.filter_by(loja_id=loja_id, status="aberto").first()


def _resumo_sessao(sessao: CaixaSessao) -> dict:
    vendas = Venda.query.filter_by(caixa_sessao_id=sessao.id, status="concluida").all()
    por_forma: dict[str, float] = {}
    for venda in vendas:
        por_forma[venda.forma_pagamento] = por_forma.get(venda.forma_pagamento, 0.0) + venda.total()

    movimentacoes = CaixaMovimentacao.query.filter_by(caixa_sessao_id=sessao.id).all()
    suprimentos = sum(float(m.valor) for m in movimentacoes if m.tipo == "suprimento")
    sangrias = sum(float(m.valor) for m in movimentacoes if m.tipo == "sangria")

    vendas_dinheiro = por_forma.get("dinheiro", 0.0)
    valor_esperado_dinheiro = float(sessao.valor_abertura) + vendas_dinheiro + suprimentos - sangrias

    return {
        "vendas_por_forma_pagamento": por_forma,
        "total_vendas": round(sum(por_forma.values()), 2),
        "quantidade_vendas": len(vendas),
        "suprimentos": round(suprimentos, 2),
        "sangrias": round(sangrias, 2),
        "valor_esperado_dinheiro": round(valor_esperado_dinheiro, 2),
    }


@caixa_bp.get("/atual")
@perfis_permitidos(*_PDV)
def caixa_atual():
    sessao = _sessao_aberta(loja_atual_id())
    if not sessao:
        return jsonify(None)
    dados = sessao.to_dict()
    dados["resumo"] = _resumo_sessao(sessao)
    return jsonify(dados)


@caixa_bp.get("/sessoes")
@perfis_permitidos(*_PDV)
def listar_sessoes():
    sessoes = query_tenant(CaixaSessao).order_by(CaixaSessao.aberto_em.desc()).limit(50).all()
    return jsonify([s.to_dict() for s in sessoes])


@caixa_bp.get("/sessoes/<int:sessao_id>")
@perfis_permitidos(*_PDV)
def obter_sessao(sessao_id):
    sessao = query_tenant(CaixaSessao).filter_by(id=sessao_id).first_or_404()
    dados = sessao.to_dict()
    dados["resumo"] = _resumo_sessao(sessao)
    return jsonify(dados)


@caixa_bp.post("/abrir")
@perfis_permitidos(*_PDV)
def abrir_caixa():
    loja_id = loja_atual_id()
    if _sessao_aberta(loja_id):
        return jsonify({"erro": "já existe um caixa aberto"}), 400

    dados = request.get_json() or {}
    sessao = CaixaSessao(
        loja_id=loja_id,
        usuario_abertura_id=get_jwt_identity(),
        valor_abertura=dados.get("valor_abertura", 0),
    )
    db.session.add(sessao)
    db.session.commit()
    return jsonify(sessao.to_dict()), 201


@caixa_bp.put("/fechar")
@perfis_permitidos(*_PDV)
def fechar_caixa():
    sessao = _sessao_aberta(loja_atual_id())
    if not sessao:
        return jsonify({"erro": "não há caixa aberto"}), 400

    dados = request.get_json() or {}
    sessao.status = "fechado"
    sessao.valor_fechamento_informado = dados.get("valor_fechamento_informado")
    sessao.observacoes = dados.get("observacoes")
    sessao.usuario_fechamento_id = get_jwt_identity()
    sessao.fechado_em = datetime.now(timezone.utc)
    db.session.commit()

    resultado = sessao.to_dict()
    resultado["resumo"] = _resumo_sessao(sessao)
    return jsonify(resultado)


@caixa_bp.get("/movimentacoes")
@perfis_permitidos(*_PDV)
def listar_movimentacoes():
    sessao = _sessao_aberta(loja_atual_id())
    if not sessao:
        return jsonify([])
    movimentacoes = (
        CaixaMovimentacao.query.filter_by(caixa_sessao_id=sessao.id).order_by(CaixaMovimentacao.criado_em.desc()).all()
    )
    return jsonify([m.to_dict() for m in movimentacoes])


@caixa_bp.post("/movimentacoes")
@perfis_permitidos(*_PDV)
def criar_movimentacao():
    loja_id = loja_atual_id()
    sessao = _sessao_aberta(loja_id)
    if not sessao:
        return jsonify({"erro": "abra o caixa antes de registrar sangria/suprimento"}), 400

    dados = request.get_json() or {}
    tipo = dados.get("tipo")
    valor = dados.get("valor")
    if tipo not in TIPOS_MOVIMENTACAO_CAIXA_VALIDOS:
        return jsonify({"erro": f"tipo inválido. Use um de: {', '.join(TIPOS_MOVIMENTACAO_CAIXA_VALIDOS)}"}), 400
    if not valor or float(valor) <= 0:
        return jsonify({"erro": "valor deve ser maior que zero"}), 400

    movimentacao = CaixaMovimentacao(
        loja_id=loja_id,
        caixa_sessao_id=sessao.id,
        usuario_id=get_jwt_identity(),
        tipo=tipo,
        valor=valor,
        motivo=dados.get("motivo"),
    )
    db.session.add(movimentacao)
    db.session.commit()
    return jsonify(movimentacao.to_dict()), 201
